crypto/s2n_tls13_keys.c

```diff
@@ -0,0 +1,201 @@
+ * [ ] exporter_master_secret
+ * [ ] resumption_master_secret
+ *
+ * The TLS 1.3 key generation can be divided into 3 phases
+ * 1. early secrets
+ * 2. handshake secrets
+ * 3. master secrets
+ *
+ * In each phase, secrets are first extracted with HKDF-Extract that takes in
+ * both an ikm (input keying material) and a salt. Some keys can be derived/expanded
+ * from the extract before a "tls13 derived" Derive-Secret is used to
+ * derive the input salt for the next phase.
+ */
+
+#define S2N_BLOB_LABEL(name, str) \
+    static uint8_t name##_data[] = str;   \
+    const struct s2n_blob name = { .data = name##_data, .size = sizeof(name##_data) - 1 };
+
+/*
+ * Define TLS 1.3 HKDF labels as specified in
+ * https://tools.ietf.org/html/rfc8446#section-7.1
+ */
+
+S2N_BLOB_LABEL(s2n_tls13_label_derived, "derived")
+
+S2N_BLOB_LABEL(s2n_tls13_label_client_early_traffic, "c e traffic")
+S2N_BLOB_LABEL(s2n_tls13_label_early_exporter_master_secret, "e exp master")
+
+S2N_BLOB_LABEL(s2n_tls13_label_client_handshake_traffic, "c hs traffic")
+S2N_BLOB_LABEL(s2n_tls13_label_server_handshake_traffic, "s hs traffic")
+
+S2N_BLOB_LABEL(s2n_tls13_label_client_application_traffic, "c ap traffic")
+S2N_BLOB_LABEL(s2n_tls13_label_server_application_traffic, "s ap traffic")
+
+S2N_BLOB_LABEL(s2n_tls13_label_exporter_master_secret, "exp master")
+S2N_BLOB_LABEL(s2n_tls13_label_resumption_master_secret, "res master")
+
+/*
+ * Traffic secret labels
+ */
+
+S2N_BLOB_LABEL(s2n_tls13_label_key, "key")
+S2N_BLOB_LABEL(s2n_tls13_label_iv, "iv")
+
+static const struct s2n_blob zero_length_blob = { .data = NULL, .size = 0 };
+
+/* Message transcript hash based on selected HMAC algorithm */
+static int s2n_tls13_message_hash(struct s2n_tls13_keys *keys, const struct s2n_blob *message, struct s2n_blob *message_digest)
+{
+    struct s2n_hash_state hash_state;
+    GUARD(s2n_hash_new(&hash_state));
+    GUARD(s2n_hash_init(&hash_state, keys->hash_algorithm));
+    GUARD(s2n_hash_update(&hash_state, message->data, message->size));
+    GUARD(s2n_hash_digest(&hash_state, message_digest->data, message_digest->size));
+    GUARD(s2n_hash_free(&hash_state));
+
+    return 0;
+}
+
+/*
+ * Initalizes the tls13_keys struct
+ */
+int s2n_tls13_keys_init(struct s2n_tls13_keys *handshake, s2n_hmac_algorithm alg)
+{
+    handshake->hmac_algorithm = alg;
+    GUARD(s2n_hmac_hash_alg(alg, &handshake->hash_algorithm));
+    GUARD(s2n_hash_digest_size(handshake->hash_algorithm, &handshake->size));
+    GUARD(s2n_blob_init(&handshake->current_secret, handshake->current_secret_bytes, handshake->size));
+    GUARD(s2n_blob_init(&handshake->derive_secret, handshake->derive_secret_bytes, handshake->size));
+    GUARD(s2n_hmac_new(&handshake->hmac));
+
+    return 0;
+}
+
+/*
+ * Derives early secrets
+ */
+int s2n_tls13_derive_early_secrets(struct s2n_tls13_keys *keys)
+{
+    s2n_tls13_key_blob(empty_salt_blob, 0);
+    s2n_tls13_key_blob(psk_ikm, keys->size);
```

Comment:
Shouldn't the psk be an input to this function?

Reply:
Eventually for 0-rtt. Here in the case of 1-rtt, psk is a zerofilled n sized array.

Reply:
Do we want a comment noting that?